crates/pathfinder/src/sync/p2p/headers.rs

```diff
@@ -0,0 +1,116 @@
+use anyhow::Context;
+use pathfinder_common::{BlockCommitmentSignature, BlockHeader, BlockNumber};
+use pathfinder_storage::Storage;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+const CHUNK_SIZE: usize = 1024;
+const CHANNEL_SIZE: usize = 2;
+
+pub async fn spawn(
+    storage: Storage,
+    p2p: (),
+    max: BlockNumber,
+    min: BlockNumber,
+) -> anyhow::Result<()> {
+    let (tx_download, rx_verify) = tokio::sync::mpsc::channel(CHANNEL_SIZE);
+    let (tx_verify, rx_persist) = tokio::sync::mpsc::channel(CHANNEL_SIZE);
+
+    let jh_download = tokio::spawn(download(tx_download, p2p, max, min, CHUNK_SIZE));
+    let jh_verify = std::thread::spawn(|| verify(rx_verify, tx_verify));
+    let jh_persist = std::thread::spawn(|| persist(rx_persist, storage));
+
+    // Since all stages are connected, we can simply wait for the final stage to complete.
+    // This works even in the event of an earlier stage failing for some reason since the
+    // channel closures will cascade and cause the final stage to exit once all data in the
+    // pipeline has been processed.
+    let result = tokio::task::spawn_blocking(|| jh_persist.join())
+        .await
+        .context("Awaiting persist thread")?;
+
+    // TODO: join the rest
+    // TODO: report any errors upstream
+
+    Ok(())
+}
+
+/// Downloads batches of block headers via p2p in reverse chronological order.
+async fn download(
+    dest: Sender<Vec<(BlockHeader, BlockCommitmentSignature)>>,
+    p2p: (), // p2p source / API
+    mut max: BlockNumber,
+    min: BlockNumber,
+    chunk_size: usize,
+) -> anyhow::Result<()> {
+    // TODO: include the hash of `max` so we know we are getting what we expect.
+
+    while max > min {
+        // TODO: download next chunk of headers.
+        // TODO: verify block hash continuity
+        // max -= headers.len();
+        // TODO: if dest.send(headers).await.is_err() { break; }
+    }
+
+    Ok(())
+}
+
+/// Verifies block header's hash and signature are correct.
+fn verify(
+    mut source: Receiver<Vec<(BlockHeader, BlockCommitmentSignature)>>,
+    dest: Sender<Vec<(BlockHeader, BlockCommitmentSignature)>>,
+) -> Result<(), (BlockHeader, BlockCommitmentSignature)> {
+    while let Some(headers) = source.blocking_recv() {
```

Comment:
Maybe it's worth doing verification via rayon thread pool in case receiving the next header takes less time than verifying the current.

Reply:
Probably yes, just keeping it simple.